### PR TITLE
Jimmy/arch 432 procpidmaps parsing

### PR DESCRIPTION
**Pull request by jimmyaxod**
This adds the ability to read /proc/<PID>/maps which we need to be able to track processes memory, and migrate it efficiently.

With this PR, we can read the memory map, we can lookup entries by pathname, address range, address page.
We can also "Diff" two maps, to see which entries are new, which have been removed, as well as which memory pages have been added/removed.

We will use this functionality to periodically read /maps for a process, see which pages have been added/removed. Then we will also be checking the soft dirty flags for pages, so that we can send this changeset (added pages, removed pages, modified pages).

The `maps.go` file has an associated test `maps_test.go` which grabs the maps for the running process, then mmaps a file, and grabs the new maps.

[Linear ticket](https://linear.app/loopholelabs/issue/ARCH-432/procpidmaps-parsing)

This is a part of [Silo memory tracking for mason](https://linear.app/loopholelabs/issue/ARCH-400/silo-memory-tracking-for-mason)

### CODE REVIEW

**Review by SuperManifolds — COMMENTED** [review]
LGTM! just one minor question

**Review by SuperManifolds — APPROVED** [review]

File: pkg/storage/memory/maps.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+	case me.PermShared && !me.PermPrivate:
+		perms[3] = MapsPermShared
+	default:
+		perms[3] = MapsPermUnknown
+	}
+
+	return fmt.Sprintf("%016x-%016x %s %08x %s %s %s", me.AddrStart, me.AddrEnd, perms, me.Offset, me.Dev, me.Inode, me.Pathname)
+}
+
+const MapsPermRead = 'r'
```
[the remark this answers is not in the document]
Comment (jimmyaxod) [reply pkg/storage/memory/maps.go:66]: Good call! bad0b2ff668dc10a0f75108790dbfba262c4be3a

File: pkg/storage/memory/maps.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+	PermExecute bool
+	PermShared  bool
+	PermPrivate bool
+	Offset      uint64
+	Dev         string
+	Inode       string
+	Pathname    string
+}
+
+func (me *MapsEntry) Equal(me2 *MapsEntry) bool {
```
[the remark this answers is not in the document]
Comment (jimmyaxod) [reply pkg/storage/memory/maps.go]: Good call! Have removed the Equals func which is great.
My only gripe (footgun) is how you can accidentally be comparing pointer values instead of struct contents with a single missing character (*).
Definitely worth it to make the code simpler and more readable (and maintainable if more fields get added etc)

Fixed in 710d4b8df92aa2210689076af6c0943133209231

File: pkg/storage/memory/maps.go [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	lines := strings.Split(string(maps), "\n")
+	for _, line := range lines {
+		// Read the header...
+		headerData := strings.Fields(line)
+		// eg
+		// 7b0bcb38c000-7b0bcb39e000 rw-p 00279000 07:00 8261                       /snap/code/206/usr/share/code/libffmpeg.so
+		entry := &MapsEntry{}
+
+		// Parse address
+		if len(headerData) > 0 {
```
[the remark this answers is not in the document]
Comment (jimmyaxod) [reply pkg/storage/memory/maps.go]: I've renamed and added some comment around this. (There is no header in the file, the var was misnamed, each line is parsed in isolation).

I've kept things fairly loose, because until we test this in the real world some against various kernels etc, we don't know exactly what we may find. We might find extra permission flags, which we can safely ignore for example, or extra fields at the end.

Specifically:
* Address range must be present, in hex.
* All other fields are optional.
* Permissions can be in any order, and any unknown permissions are ignored.
* Offset must be valid hex.
* Any fields past the pathname are ignored.

53311a1fab60508df4a9c55846559aa2f44db183

Would you change the parsing restrictions? Alternative would be to collect a variety of examples from kernels we want to support, then we can assert they are readable.
Asserting that parsing garbage returns an error is harder - it's always a question of what sort of garbage - maybe it's completely correct but has an extra space in an unexpected place.... But I think we find those edge cases in the wild, and add them as tests in later PR if needed. WDYT?
